File: 02.double linked list/python/double linked list.py
class Node:
    def __init__(self, v):
        self.value = v
        self.prev = None
        self.next = None


class LinkedList2:
    def __init__(self):
        self.head = None
        self.tail = None

    def add_in_tail(self, item):
        if self.head is None:
            self.head = item
            item.prev = None
            item.next = None
        else:
            self.tail.next = item
            item.prev = self.tail
        self.tail = item

    def find(self, val):
        node = self.head
        while node is not None:
            if node.value == val:
                return node
            node = node.next
        return None

    def find_all(self, val):
        node = self.head
        nodes = []
        while node is not None:
            if node.value == val:
                nodes.append(node)
            node = node.next
        return nodes

    def delete(self, val, all=False):
        node = self.head

        while node is not None and node.value == val:  # Удаление если это первый(е) элемент(ы)
            self.head = self.head.next
            if self.head is None:
                self.tail = None
            else:
                self.head.prev = None
            node = self.head
            if not all:
                return

        before = None

        while node is not None:
            while node is not None and node.value != val:
                before = node
                node = node.next

            if node is None:
                return

            before.next = node.next
            node = node.next
            if node is None:
                self.tail = before
            else:
                node.prev = before

            if not all:
                return

    def clean(self):
        current = self.head
        pointer = self.head.next
        while pointer:
            current.next = None
            current.prev = None
            current = pointer
            pointer = pointer.next
        self.head = None
        self.tail = None

    def len(self):
        counter = 0
        node = self.head
        while node:
            counter += 1
            node = node.next
        return counter

    def insert(self, afterNode, newNode):
        node = self.head

        if afterNode is None:  # Вставка в начало
            self.add_in_tail(newNode)
            return

        if self.tail == afterNode:  # Вставка в конец
            self.add_in_tail(newNode)
            return

        if node is None:
            return

        while node.next is not None:  # Вставка в середину
            if node == afterNode:
                newNode.next = node.next
                newNode.prev = node
                node.next.prev = newNode
                node.next = newNode
                return
            node = node.next

    def add_in_head(self, newNode):
        newNode.prev = None
        newNode.next = self.head
        if self.head:
            self.head.prev = newNode
        else:
            self.tail = newNode
        self.head = newNode
        return

    def traverse(self):
        tmp = None
        node = self.tail

        while node is not None:
            tmp = node.next
            node.next = node.prev
            node.prev = tmp
            node = node.next

        if tmp is not None:
            self.head, self.tail = self.tail, self.head

File: 02.double linked list/golang/double_linked_list.go
package main

import (
	"errors"
	"os"
	"reflect"
)

type Node struct {
	prev  *Node
	next  *Node
	value int
}

type LinkedList2 struct {
	head *Node
	tail *Node
}

func (l *LinkedList2) AddInTail(item Node) {
	if l.head == nil {
		l.head = &item
		l.head.next = nil
		l.head.prev = nil
	} else {
		l.tail.next = &item
		item.prev = l.tail
		item.next = nil
	}
	l.tail = &item
}

func (l *LinkedList2) Count() int {
	count := 0
	node := l.head
	for node != nil {
		count++
		node = node.next
	}
	return count
}

// error не nil, если узел не найден
func (l *LinkedList2) Find(n int) (Node, error) {
	node := l.head
	for node != nil {
		if node.value == n {
			return *node, nil
		}
		node = node.next
	}
	return Node{value: -1, next: nil}, errors.New("node not found")
}

func (l *LinkedList2) FindAll(n int) []Node {
	var nodes []Node
	node := l.head
	for node != nil {
		if node.value == n {
			nodes = append(nodes, *node)
		}
		node = node.next
	}
	return nodes
}

func (l *LinkedList2) Delete(n int, all bool) {
	if all {
		for l.head != nil && l.head.value == n {
			l.head = l.head.next
			if l.head == nil {
				l.tail = nil
			} else {
				l.head.prev = nil
			}
		}
		if l.head == nil {
			return
		}
		node := l.head
		for node != nil {
			if node.next != nil && node.next.value == n {
				toRemove := node.next
				node.next = toRemove.next
				if node.next == nil {
					l.tail = node
				} else {
					node.next.prev = node
				}
			} else {
				node = node.next
			}
		}
	} else {
		if l.head == nil {
			return
		}
		if l.head.value == n {
			l.head = l.head.next
			if l.head == nil {
				l.tail = nil
			} else {
				l.head.prev = nil
			}
			return
		}
		node := l.head
		for node.next != nil {
			if node.next.value == n {
				toRemove := node.next
				node.next = toRemove.next
				if node.next == nil {
					l.tail = node
				} else {
					node.next.prev = node
				}
				return
			}
			node = node.next
		}
	}
}

func (l *LinkedList2) Insert(after *Node, add Node) {
	if after == nil {
		add.next = l.head
		add.prev = nil
		if l.head != nil {
			l.head.prev = &add
		} else {
			l.tail = &add
		}
		l.head = &add
		return
	}
	node := l.head
	for node != nil {
		if node == after {
			add.next = node.next
			add.prev = node
			if node.next != nil {
				node.next.prev = &add
			} else {
				l.tail = &add
			}
			node.next = &add
			return
		}
		node = node.next
	}
}

func (l *LinkedList2) InsertFirst(first Node) {
	l.Insert(nil, first)
}

func (l *LinkedList2) Clean() {
	current := l.head
	for current != nil {
		next := current.next
		current.next = nil
		current.prev = nil
		current = next
	}
	l.head = nil
	l.tail = nil
}
